jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/profiles/ProfileUtils.kt

```diff
@@ -0,0 +1,52 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.core.credentials.profiles
+
+import com.intellij.util.text.nullize
+import software.amazon.awssdk.profiles.Profile
+import software.amazon.awssdk.profiles.ProfileProperty
+import software.aws.toolkits.resources.message
+
+fun traverseCredentialChain(profiles: Map<String, Profile>, topLevelProfile: Profile): Sequence<Profile> = sequence {
+    val profileChain = linkedSetOf<String>()
+    var currentProfile = topLevelProfile
+
+    yield(currentProfile)
+
+    while (currentProfile.propertyExists(ProfileProperty.ROLE_ARN)) {
+        val currentProfileName = currentProfile.name()
+        if (!profileChain.add(currentProfileName)) {
+            val chain = profileChain.joinToString("->", postfix = "->$currentProfileName")
+            throw IllegalArgumentException(message("credentials.profile.circular_profiles", topLevelProfile.name(), chain))
+        }
+
+        val sourceProfile = currentProfile.requiredProperty(ProfileProperty.SOURCE_PROFILE)
```

Comment:
So we don't support `credential_source` chains? 

From https://docs.aws.amazon.com/cli/latest/topic/config-vars.html
> credential_source - The credential provider to use to get credentials for the initial assume-role call. This 
  parameter cannot be provided alongside source_profile. Valid values are:
  Environment to pull source credentials from environment variables.
  Ec2InstanceMetadata to use the EC2 instance role as source credentials.
  EcsContainer to use the ECS container credentials as the source credentials.

Reply:
Nope, we don't today since we don't support any of the other providers.